File: LiDAR2BIM-Registration/preprocess/geometry/point_cloud.py
import numpy as np
from preprocess.geometry.lineseg import LineSegment
from preprocess.geometry.corner import Corner
from preprocess.geometry.pc2img import Points2Image
from preprocess.utils import stats
from itertools import combinations

import open3d as o3d
from scipy.spatial.transform import Rotation as R
from copy import deepcopy
from typing import List
import gc

class PointCloudManager:
    def __init__(self, points: np.ndarray, cfg):
        self.cfg = cfg
        self.points = points

        self.png = None
        self.T_xyz2hw = None
        self.T_hw2xyz = None
        self.T_hw2xy = None

    def extract_hash_descriptors(self, corners: List[Corner]):
        linesegments = []
        if self.cfg.desc_name == "triangle":
            pass
        elif self.cfg.desc_name == "linesegment":
            linesegments = [
                LineSegment().from_corners(corners[i], corners[j])
                for i, j in combinations(range(len(corners)), 2)
            ]
        elif self.cfg.desc_name == "TriDescV2":
            pass
        elif self.cfg.desc_name == "TriDescV3":
            pass
        elif self.cfg.desc_name == "TriDescGr6D":
            pass
        elif self.cfg.desc_name == "TriDescGr7D":
            pass
        elif self.cfg.desc_name == "TriDescGr9D":
            pass
        stats.add("corners", len(corners))
        stats.add("linesegments", len(linesegments))
        geom_sets = self.desc_hashtable.build_geom_set(corners, linesegments, radius=self.cfg.max_side_length)
        stats.add("geom_sets", len(geom_sets))
        # Build hash table
        self.desc_hashtable.build_hash_table(geom_sets)
        return True

    def get_img(self):
        pc2img = Points2Image(self.points, self.cfg)
        self.png = pc2img.get_png()
        self.T_xyz2hw = pc2img.T_xyz2hw
        self.T_hw2xyz = pc2img.T_hw2xyz
        self.T_hw2xy = pc2img.T_hw2xy


class PCD2d:
    def __init__(
        self,
        walls_pcd: o3d.geometry.PointCloud,
        ground_pcd: o3d.geometry.PointCloud,
        cfg,
    ):
        self.walls_pcd = walls_pcd
        self.ground_pcd = ground_pcd
        self.cfg = cfg
        self.ref = o3d.geometry.TriangleMesh.create_coordinate_frame(
            size=55, origin=[0, 0, 0]
        )
        self.upper_bound = cfg.pcd2d.max_height
        self.lower_bound = cfg.pcd2d.min_height
        self.voxel_size = cfg.voxel

        self.T_rp = np.eye(4)
        self.T_rp_inv = np.eye(4)
        self.og_walls = None
        self.og_ground = None
        self.og_walls_cropped = None
        self.walls_2d = None

    # @profile
    def get_pcd2d(self, lidar_origin="default"):
        # reset rp to 0 to align gravity
        if not self.reset_rp(lidar_origin=lidar_origin) or len(self.walls_pcd.points) == 0:  # rpy and translation
            return False
        # crop the walls
        self.crop_walls()
        # self.viz_crop_walls()

        # flatten the walls
        self.flatten_walls()
        return True

    def clear_cache(self):
        refs = gc.get_referrers(self.ground_pcd)
        print(f"Number of referrers: {len(refs)}")

        # 输出每个引用对象的信息
        for i, ref in enumerate(refs):
            print(f"Referrer {i}: {ref}")
            print(f"Type: {type(ref)}")
            if isinstance(ref, dict):
                print(f"Keys: {list(ref.keys())[:10]}")  # 只打印前10个键
            elif isinstance(ref, list):
                print(f"First 10 elements: {ref[:10]}")  # 只打印前10个元素
            elif isinstance(ref, tuple):
                print(f"First 10 elements: {ref[:10]}")  # 只打印前10个元素
            print('-' * 80)

        del self.og_walls
        del self.og_ground
        del self.og_walls_cropped
        del self.walls_2d
        del self.walls_pcd
        del self.ground_pcd

        gc.collect()

    def reset_rp(self, lidar_origin="default") -> np.ndarray:
        # timer.tic("fit_plane")
        down_ground = self.ground_pcd.voxel_down_sample(voxel_size=1.0)
        if len(down_ground.points) < 3:
            print("No enough points to fit the plane.")
            return False
        # plane = fit_plane_PCA(np.array(down_ground.points))
        # timer.toc("fit_plane")
        plane = down_ground.segment_plane(0.1, 3, 20)[0]
        t_l2w = np.array([0, 0, 0])  # lidar to world
        if lidar_origin == "default":
            t_l2w = proj_pt2pl(t_l2w, plane)
        if lidar_origin == "min":
            wall_min = np.min(np.array(self.walls_pcd.points), axis=0)
            wm_on_plane = proj_pt2pl(wall_min, plane)
            t_l2w = wm_on_plane

        v1 = np.array([0, 0, 1])
        v2 = np.array(plane[:3])
        R = R_from_2vecs(v1, v2)
        t = -np.dot(R.T, t_l2w)
        T = np.eye(4)
        T[:3, :3] = R.T
        T[:3, 3] = t

        self.T_rp = T
        self.T_rp_inv = np.linalg.inv(T)
        self.og_walls = deepcopy(self.walls_pcd).transform(self.T_rp)
        self.og_ground = deepcopy(self.ground_pcd).transform(self.T_rp)
        return True

    def crop_walls(self):
        # get the crop box
        bound_pcd = get_bound_pcd(self.og_walls)
        crop_box = get_crop_box(bound_pcd, self.upper_bound, self.lower_bound)
        # crop the walls
        self.og_walls_cropped = self.og_walls.crop(crop_box)
    def viz_crop_walls(self):
        o3d.visualization.draw_geometries(
            [
                self.og_ground,
                deepcopy(self.og_walls_cropped).paint_uniform_color([1, 0, 0]),
                self.ref,
            ]
        )

    def flatten_walls(self):
        # set wall z to 0
        xyz = np.array(self.og_walls_cropped.points)
        xyz[:, 2] = 0
        walls_2d = o3d.geometry.PointCloud()
        walls_2d.points = o3d.utility.Vector3dVector(xyz)
        walls_2d.colors = self.og_walls_cropped.colors
        walls_2d = walls_2d.voxel_down_sample(voxel_size=self.voxel_size)
        self.walls_2d = walls_2d



def R_from_2vecs(v1, v2) -> np.ndarray:
    # v1, v2: 3x1 np.array
    v1 = v1 / np.linalg.norm(v1)
    v2 = v2 / np.linalg.norm(v2)
    axis = np.cross(v1, v2)
    axis_length = np.linalg.norm(axis)
    angle = np.arccos(np.dot(v1, v2))
    return R.from_rotvec(axis / axis_length * angle).as_matrix()


def fit_plane_PCA(points):
    centroid = np.mean(points, axis=0)
    points_centered = points - centroid

    H = np.dot(points_centered.T, points_centered)

    eigenvalues, eigenvectors = np.linalg.eigh(H)

    normal = eigenvectors[:, np.argmin(eigenvalues)]
    d = -np.dot(normal, centroid)
    plane = np.hstack([normal, d])

    return plane


def proj_pt2line(pt, mean_pt, v):
    t = np.dot(pt - mean_pt, v) / np.dot(v, v)
    return mean_pt + t * v


def get_pt2pl_dist(pt: np.ndarray, plane) -> float:
    a, b, c, d = plane
    return a * pt[0] + b * pt[1] + c * pt[2] + d / np.sqrt(a**2 + b**2 + c**2)


def proj_pt2pl(pt: np.ndarray, plane) -> np.ndarray:
    a, b, c, d = plane
    x, y, z = pt
    pt2pl_dist = get_pt2pl_dist(pt, plane)
    norm = np.sqrt(a**2 + b**2 + c**2)
    return np.array(
        [
            x - a * pt2pl_dist / norm,
            y - b * pt2pl_dist / norm,
            z - c * pt2pl_dist / norm,
        ]
    )


def get_crop_box(bound_pcd, upper_bound, lower_bound):
    bound_pts = np.array(bound_pcd.points)
    # set z to 0
    bound_pts[:, 2] = 0
    normal = np.array([0, 0, 1])

    # get the min and max of the bound_pts_on_plane
    min_bound, max_bound = get_min_max(bound_pts)
    min_bound = min_bound + normal * lower_bound
    max_bound = max_bound + normal * upper_bound

    crop_box = o3d.geometry.AxisAlignedBoundingBox(min_bound, max_bound)
    crop_box.color = [0, 1, 0]
    return crop_box


def get_bound_pcd(pcd) -> o3d.geometry.PointCloud:
    aabb = pcd.get_axis_aligned_bounding_box()
    min_bound = aabb.get_min_bound()
    max_bound = aabb.get_max_bound()
    bound_pcd = o3d.geometry.PointCloud()
    bound_pcd.points = o3d.utility.Vector3dVector(
        np.array(
            [
                min_bound,
                [max_bound[0], min_bound[1], min_bound[2]],
                [max_bound[0], max_bound[1], min_bound[2]],
                [min_bound[0], max_bound[1], min_bound[2]],
                min_bound,
            ]
        )
    )
    return bound_pcd


def get_min_max(bound_pts: np.ndarray) -> tuple:
    min_x = np.min(bound_pts[:, 0])
    max_x = np.max(bound_pts[:, 0])
    min_y = np.min(bound_pts[:, 1])
    max_y = np.max(bound_pts[:, 1])

    # the point should be on the plane
    min_z = 0
    max_z = 0

    min_bound = np.array([min_x, min_y, min_z])
    max_bound = np.array([max_x, max_y, max_z])

    return min_bound, max_bound

File: LiDAR2BIM-Registration/src/frontend/geometry/point_cloud.cpp
#include "frontend/geometry/point_cloud.h"
#include <iostream>
#include <Eigen/Dense>
#include <memory>

PCD2d::PCD2d() : wallsPcd_(), groundPcd_(),
				 upperBound_(cfg::maxHeight), lowerBound_(cfg::minHeight),
				 voxelSize_(cfg::voxel_size),
				 T_rp_(Eigen::Matrix4d::Identity()),
				 T_rp_inv_(Eigen::Matrix4d::Identity()), ogWalls_(),
				 ogGround_(){};

PCD2d::PCD2d(
		pcl::PointCloud<pcl::PointXYZ>::Ptr groundPcd
) : wallsPcd_(), groundPcd_(groundPcd),
	upperBound_(cfg::maxHeight), lowerBound_(cfg::minHeight),
	voxelSize_(cfg::voxel_size),
	T_rp_(Eigen::Matrix4d::Identity()),
	T_rp_inv_(Eigen::Matrix4d::Identity()), ogWalls_(),
	ogGround_(groundPcd) {

//	pcl::VoxelGrid<PointT> voxelFilter;
//	voxelFilter.setInputCloud(groundPcd_);
//	voxelFilter.setLeafSize(1.0f, 1.0f, 1.0f); // 设置体素大小
//	voxelFilter.filter(*groundPcd_);
}

PCD2d::PCD2d(
		pcl::PointCloud<pcl::PointXYZ>::Ptr wallsPcd,
		pcl::PointCloud<pcl::PointXYZ>::Ptr groundPcd
) : wallsPcd_(wallsPcd), groundPcd_(groundPcd),
	upperBound_(cfg::maxHeight), lowerBound_(cfg::minHeight),
	voxelSize_(cfg::voxel_size), T_rp_(Eigen::Matrix4d::Identity()),
	T_rp_inv_(Eigen::Matrix4d::Identity()), ogWalls_(wallsPcd), ogGround_(groundPcd) {
	
	pcl::VoxelGrid<PointT> voxelFilter;
	voxelFilter.setInputCloud(groundPcd_);
	voxelFilter.setLeafSize(1.0f, 1.0f, 1.0f); // 设置体素大小
	voxelFilter.filter(*groundPcd_);
}

bool PCD2d::GetPcd2d(const std::string &lidarOrigin) {
	if (!ResetRP(lidarOrigin)) return false;
//	CropWalls();
//	FlattenWalls();
	return true;
}

bool PCD2d::ResetRP(const std::string &lidarOrigin) {
//	auto downGround = groundPcd_->VoxelDownSample(1.0);
//	if (downGround->points_.size() < 3) {
//		std::cout << "Not enough points to fit the plane." << std::endl;
//		return false;
//	}
//	auto plane = downGround->SegmentPlane(0.1, 3, 20).first;
	timer.tic("Reset RP");
	
	
	if (groundPcd_->points.size() < 3) {
		std::cout << "Not enough points to fit the plane." << std::endl;
		return false;
	}
	pcl::ModelCoefficients::Ptr planeCoefficients(new pcl::ModelCoefficients);
	pcl::PointIndices::Ptr inliers(new pcl::PointIndices);
	
	pcl::SACSegmentation<PointT> seg;
	seg.setOptimizeCoefficients(true);
	seg.setModelType(pcl::SACMODEL_PLANE);
	seg.setMethodType(pcl::SAC_RANSAC);
	seg.setMaxIterations(1000);
	seg.setDistanceThreshold(0.1);
	seg.setInputCloud(groundPcd_);
	seg.segment(*inliers, *planeCoefficients);
	if (inliers->indices.size() == 0) {
		std::cerr << "Could not estimate a planar model for the given dataset." << std::endl;
		return false;
	}
	Eigen::Vector4d plane;
	plane << planeCoefficients->values[0], planeCoefficients->values[1],
			planeCoefficients->values[2], planeCoefficients->values[3];
	Eigen::Vector3d t_l2w(0, 0, 0);
	
	t_l2w = ProjPt2Pl(t_l2w, plane);
	
	
	Eigen::Vector3d v1(0, 0, 1);
	Eigen::Vector3d v2(plane.head<3>());
	Eigen::Matrix3d R = RFrom2Vecs(v1, v2);
	Eigen::Vector3d t = -R.transpose() * t_l2w;
	
	T_rp_.block<3, 3>(0, 0) = R.transpose();
	T_rp_.block<3, 1>(0, 3) = t;
	
	T_rp_inv_ = T_rp_.inverse();
	//transform walls and ground
//	Eigen::Matrix4f T_rp_float = T_rp_.cast<float>();
//	pcl::transformPointCloud(*ogWalls_, *ogWalls_, T_rp_);
//	pcl::transformPointCloud(*ogGround_, *ogGround_, T_rp_);
	timer.toc("Reset RP");
	return true;
}

void PCD2d::CropWalls() {
	pcl::CropBox<PointT> cropBox = GetCropBox(GetBoundPcd(ogWalls_), upperBound_, lowerBound_);
	cropBox.setInputCloud(ogWalls_);
	cropBox.filter(*ogWallsCropped_);
}

void PCD2d::FlattenWalls() {
	pcl::PointCloud<PointT>::Ptr tempCloud(new pcl::PointCloud<PointT>(*ogWallsCropped_));
	for (auto &point: tempCloud->points) {
		point.z = 0;
	}
	pcl::VoxelGrid<PointT> voxelGrid;
	voxelGrid.setInputCloud(tempCloud);
	voxelGrid.setLeafSize(voxelSize_, voxelSize_, voxelSize_);
	voxelGrid.filter(*walls2d_);
}

Eigen::Matrix3d PCD2d::RFrom2Vecs(const Eigen::Vector3d &v1, const Eigen::Vector3d &v2) {
	Eigen::Vector3d v1Norm = v1.normalized();
	Eigen::Vector3d v2Norm = v2.normalized();
	Eigen::Vector3d axis = v1Norm.cross(v2Norm);
	double axisLength = axis.norm();
	double angle = acos(v1Norm.dot(v2Norm));
	return Eigen::AngleAxisd(angle, axis / axisLength).toRotationMatrix();
}


double PCD2d::GetPt2PlDist(const Eigen::Vector3d &pt, const Eigen::Vector4d &plane) {
	double a = plane[0], b = plane[1], c = plane[2], d = plane[3];
	return (a * pt[0] + b * pt[1] + c * pt[2] + d) / std::sqrt(a * a + b * b + c * c);
}

Eigen::Vector3d PCD2d::ProjPt2Pl(const Eigen::Vector3d &pt, const Eigen::Vector4d &plane) {
	double a = plane[0], b = plane[1], c = plane[2], d = plane[3];
	double x = pt[0], y = pt[1], z = pt[2];
	double pt2pl_dist = GetPt2PlDist(pt, plane);
	double norm = std::sqrt(a * a + b * b + c * c);
	return Eigen::Vector3d(
			x - a * pt2pl_dist / norm,
			y - b * pt2pl_dist / norm,
			z - c * pt2pl_dist / norm
	);
}

pcl::PointCloud<PointT>::Ptr PCD2d::GetBoundPcd(const pcl::PointCloud<PointT>::Ptr &pcd) {
	Eigen::Vector4f minPt, maxPt;
	pcl::getMinMax3D(*pcd, minPt, maxPt);
	
	pcl::PointCloud<PointT>::Ptr boundPcd(new pcl::PointCloud<PointT>);
	boundPcd->emplace_back(minPt.x(), minPt.y(), minPt.z());
	boundPcd->emplace_back(maxPt.x(), minPt.y(), minPt.z());
	boundPcd->emplace_back(maxPt.x(), maxPt.y(), minPt.z());
	boundPcd->emplace_back(minPt.x(), maxPt.y(), minPt.z());
	
	return boundPcd;
}

std::pair<Eigen::Vector3f, Eigen::Vector3f> PCD2d::GetMinMax(const pcl::PointCloud<PointT>::Ptr &boundPcd) {
	Eigen::Vector4f minPt, maxPt;
	pcl::getMinMax3D(*boundPcd, minPt, maxPt);
	
	Eigen::Vector3f minBound(minPt.x(), minPt.y(), 0.0f);
	Eigen::Vector3f maxBound(maxPt.x(), maxPt.y(), 0.0f);
	
	return {minBound, maxBound};
}

pcl::CropBox<PointT>
PCD2d::GetCropBox(const pcl::PointCloud<PointT>::Ptr &boundPcd, float upperBound, float lowerBound) {
	auto [minBound, maxBound] = GetMinMax(boundPcd);
	
	minBound.z() = lowerBound;
	maxBound.z() = upperBound;
	
	pcl::CropBox<PointT> cropBox;
	cropBox.setMin(Eigen::Vector4f(minBound.x(), minBound.y(), minBound.z(), 1.0));
	cropBox.setMax(Eigen::Vector4f(maxBound.x(), maxBound.y(), maxBound.z(), 1.0));
	
	return cropBox;
}